proxy.go

```diff
@@ -367,6 +359,20 @@ func (rp *reverseProxy) serveFromCache(s *scope, srw *statResponseWriter, req *h
 	}
 }
 
+func (rp *reverseProxy) completeTransaction(s *scope, statusCode int, userCache *cache.AsyncCache, key *cache.Key, q []byte) {
+	if _, ok := clickhouseNonRecoverableStatusCodes[statusCode]; ok {
+		if err := userCache.Fail(key); err != nil {
+			log.Errorf("%s: %s; query: %q", s, err, q)
+		}
+	} else {
+		if err := userCache.Complete(key); err != nil {
+			log.Errorf("%s: %s; query: %q", s, err, q)
+		}
+	}
+}
+
+var clickhouseNonRecoverableStatusCodes = map[int]struct{}{400: {}, 404: {}, 500: {}}
```

Comment:
cf our discussion on slack. There is at least the 408 code that is missing and is very important. And maybe doing a list of clickhouseRecoverableStatusCodes instead of clickhouseNonRecoverableStatusCodes might be safer